vyper/semantics/analysis/local.py

```diff
@@ -551,62 +516,190 @@ def visit_Log(self, node):
-        self.visit(node.left)  # type: ignore[attr-defined]
-        self.visit(node.right)  # type: ignore[attr-defined]
-
-    def visit_Dict(self, node: vy_ast.Dict) -> None:
-        for key in node.keys:
-            self.visit(key)
+        if self.func.mutability is not StateMutability.PAYABLE:
+            _validate_msg_value_access(node)
+
+        if self.func.mutability == StateMutability.PURE:
+            _validate_pure_access(node, type_)
+
+        value_type = get_exact_type_from_node(node.value)
+        _validate_address_code_attribute(node, value_type)
+
+        self.visit(node.value, value_type)
+
+    def visit_BinOp(self, node: vy_ast.BinOp, type_: VyperType) -> None:
+        self.visit(node.left, type_)
+        self.visit(node.right, type_)
+
+    def visit_BoolOp(self, node: vy_ast.BoolOp, type_: VyperType) -> None:
         for value in node.values:
-            self.visit(value)
+            self.visit(value, BoolT())
+
+    def visit_Call(self, node: vy_ast.Call, type_: VyperType) -> None:
+        call_type = get_exact_type_from_node(node.func)
+        self.visit(node.func, call_type)
+
+        if isinstance(call_type, ContractFunctionT):
+            # function calls
+            if call_type.is_internal:
+                self.func.called_functions.add(call_type)
+            for arg, typ in zip(node.args, call_type.argument_types):
+                self.visit(arg, typ)
+            for kwarg in node.keywords:
+                # We should only see special kwargs
+                self.visit(kwarg.value, call_type.call_site_kwargs[kwarg.arg].typ)
+
+        elif is_type_t(call_type, EventT):
+            # events have no kwargs
+            for arg, typ in zip(node.args, list(call_type.typedef.arguments.values())):
+                self.visit(arg, typ)
+        elif is_type_t(call_type, StructT):
+            # struct ctors
+            # ctors have no kwargs
+            for value, arg_type in zip(
+                node.args[0].values, list(call_type.typedef.members.values())
+            ):
+                self.visit(value, arg_type)
+        elif isinstance(call_type, MemberFunctionT):
+            assert len(node.args) == len(call_type.arg_types)
+            for arg, arg_type in zip(node.args, call_type.arg_types):
+                self.visit(arg, arg_type)
+        else:
+            # builtin functions
+            arg_types = call_type.infer_arg_types(node)
+            for arg, arg_type in zip(node.args, arg_types):
+                self.visit(arg, arg_type)
+            kwarg_types = call_type.infer_kwarg_types(node)
+            for kwarg in node.keywords:
+                self.visit(kwarg.value, kwarg_types[kwarg.arg])
+
+    def visit_Compare(self, node: vy_ast.Compare, type_: VyperType) -> None:
+        if isinstance(node.op, (vy_ast.In, vy_ast.NotIn)):
+            if isinstance(node.right, vy_ast.List):
+                cmp_type = get_common_types(node.left, *node.right.elements).pop()
```

Comment:
Renamed `type_` to `cmp_type` in this function because the comparison type is distinct from the boolean type of the `Compare` node.